### PR TITLE
Timeseries metric

### PR DESCRIPTION
**Pull request by rajs96**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Adds timeseries metric to the library.

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: torchmetrics/timeseries.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        super().__init__()
+        self.metric = metric
+
+    @property
+    def n_timesteps(self) -> int:
+        return len(self)
+
+    def add_timestep(self, preds: torch.Tensor, targets: torch.Tensor) -> None:
+        """Adds a metric module to the module list"""
+        module = self.metric()
```
Comment (SkafteNicki) [inline torchmetrics/timeseries.py:27]: assuming that users have passed in a instantiated metric, we probably need to do a deep copy here
```suggestion
        module = deepcopy(self.metric)
```

File: torchmetrics/timeseries.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+from torchmetrics.metric import Metric
+
+
+class TimeSeriesMetric(nn.ModuleList):
+    """
+    A class that keeps track of metrics as a timeseries, and implements useful methods.
+
+    Args:
+        metric: the torchmetric to keep track of at each timestep.
+    """
```
Comment (SkafteNicki) [inline torchmetrics/timeseries.py:15]: an example would be nice here :]

File: torchmetrics/timeseries.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    def __init__(self, metric: Metric):
+        super().__init__()
+        self.metric = metric
+
+    @property
+    def n_timesteps(self) -> int:
+        return len(self)
+
+    def add_timestep(self, preds: torch.Tensor, targets: torch.Tensor) -> None:
```
Comment (SkafteNicki) [inline torchmetrics/timeseries.py:25]: to make sure that this wrapper works with all metrics I think we need to keep it very general
```suggestion
    def add_timestep(self, *args, **kwargs) -> None:
```

File: torchmetrics/timeseries.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        """Adds a metric module to the module list"""
+        module = self.metric()
+        module.update(preds, targets)
```
Comment (SkafteNicki) [inline torchmetrics/timeseries.py:28]:
```suggestion
        module.update(*args, **kwargs)
```

File: torchmetrics/timeseries.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+        Returns:
+            The best metric value, and optionally the timestep.
+        """
+        idx, max = torch.max(self.all_metrics(), 0)
+        if return_timestep:
+            return idx.item(), max.item()
+        return max.item()
+
+    def all_metrics(self) -> torch.Tensor:
```
Comment (SkafteNicki) [inline torchmetrics/timeseries.py:46]: since this is just calling compute of all modules I feel like it is a fitting name
```suggestion
    def compute(self) -> torch.Tensor:
```

**Review by SkafteNicki — COMMENTED** [review]
Overall it looks good to me, but I am basically missing a bit of context.
Could you provide an example on how it is intended to be used

File: torchmetrics/timeseries.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    Args:
+        metric: the torchmetric to keep track of at each timestep.
+    """
+
+    def __init__(self, metric: Metric):
+        super().__init__()
+        self.metric = metric
+
+    @property
+    def n_timesteps(self) -> int:
```
Comment (Borda) [inline torchmetrics/timeseries.py:22]:
```suggestion
    def num_timesteps(self) -> int:
```
or `nb_timesteps` but is it needed if it just len? or later it can be different?

File: torchmetrics/timeseries.py (hunk 1 above)
Comment (Borda) [reply torchmetrics/timeseries.py:27]: I think that here the `TimeSeriesMetric.metric` shall me metric type not instance

**Comment by Borda** [comment]
@rajs96 mind check comments and then we can turn it to PR (not just draft)

File: torchmetrics/timeseries.py (hunk 6 above)
Comment (Borda) [reply torchmetrics/timeseries.py:22]: well, shall it rather count of all preds/targets?

File: torchmetrics/timeseries.py (hunk 1 above)
Comment (Borda) [reply torchmetrics/timeseries.py:27]: but then it won't have the past measures so it would be waiting of compute...

**Comment by Borda** [comment]
@rajs96 could you also add to the PR description add a use-case/application?
after we ll add in the docs anyway :rabbit:

**Comment by Borda** [comment]
@rajs96 how is it going here?

**Comment by SkafteNicki** [comment]
Hi @rajs96, I took some time and somewhat completed the PR. Please take a look and see if it still looks like what you proposed.

**Comment by Borda** [comment]
@rajs96 how is it going here? :rabbit:

File: tests/wrappers/test_tracker.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+            MeanAbsoluteError,
+            (
+                torch.randn(
+                    50,
+                ),
+                torch.randn(
+                    50,
+                ),
+            ),
+            False,
```
Comment (Borda) [inline tests/wrappers/test_tracker.py:76]:
```suggestion
            False
```

File: tests/wrappers/test_tracker.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+            MeanSquaredError,
+            (
+                torch.randn(
+                    50,
+                ),
+                torch.randn(
+                    50,
+                ),
+            ),
+            False,
```
Comment (Borda) [inline tests/wrappers/test_tracker.py:64]:
```suggestion
            MeanSquaredError,
            (
                torch.randn(
                    50
                ),
                torch.randn(
                    50
                ),
            ),
            False
```

File: tests/wrappers/test_tracker.py (hunk 7 above)
Comment (Borda) [inline tests/wrappers/test_tracker.py:76]:
```suggestion
            MeanAbsoluteError,
            (
                torch.randn(
                    50
                ),
                torch.randn(
                    50
                ),
            ),
            False
```

File: tests/wrappers/test_tracker.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        ),
+        (
+            MeanAbsoluteError,
+            (
+                torch.randn(50),
+                torch.randn(50),
+            ),
+            False,
+        ),
```
Comment (Borda) [inline tests/wrappers/test_tracker.py:69]:
```suggestion
        (
            MeanSquaredError,
            (
                torch.randn(50),
                torch.randn(50)
            ),
            False
        ),
        (
            MeanAbsoluteError,
            (
                torch.randn(50),
                torch.randn(50)
            ),
            False
        ),
```

File: CHANGELOG.md [hunk 10]
```diff
@@ -39,6 +39,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Added support for negative targets in `nDCG` metric ([#378](https://github.com/PyTorchLightning/metrics/pull/378))
 
 
+- Added `MetricTracker` wrapper metric for keeping track of the same metric over multiple epochs ([#238](https://github.com/PyTorchLightning/metrics/pull/238))
+
```
Comment (Borda) [inline CHANGELOG.md:43]:
```suggestion


```

File: torchmetrics/wrappers/tracker.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+        >>> from torchmetrics import Accuracy, MetricTracker
+        >>> tracker = MetricTracker(Accuracy(num_classes=10))
+        >>> for epoch in range(5):  # doctest: +SKIP
+        ...     tracker.increment()  # doctest: +SKIP
+        ...     for batch_idx in range(5):  # doctest: +SKIP
+        ...         preds, target = torch.randint(10, (100,)), torch.randint(10, (100,))  # doctest: +SKIP
+        ...         tracker.update(preds, target)  # doctest: +SKIP
+        ...     print(f"current acc={tracker.compute()}")  # doctest: +SKIP
+        >>> best_acc, which_epoch = tracker.best_metric(return_step=True)  # doctest: +SKIP
+        >>> all_values = tracker.compute_all()  # doctest: +SKIP
```
Comment (Borda) [inline torchmetrics/wrappers/tracker.py:51]:
```suggestion

        from torchmetrics import Accuracy, MetricTracker
        tracker = MetricTracker(Accuracy(num_classes=10))
        for epoch in range(5):
            tracker.increment()
            for batch_idx in range(5):
                preds, target = torch.randint(10, (100,)), torch.randint(10, (100,))
                tracker.update(preds, target)
            print(f"current acc={tracker.compute()}")
        best_acc, which_epoch = tracker.best_metric(return_step=True)
        all_values = tracker.compute_all()
```

File: tests/wrappers/test_tracker.py [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    "method, method_input",
+    [
+        ("update", (torch.randint(10, (50,)), torch.randint(10, (50,)))),
+        ("forward", (torch.randint(10, (50,)), torch.randint(10, (50,)))),
+        ("compute", None),
+    ],
+)
+def test_raises_error_if_increment_not_called(method, method_input):
+    with pytest.raises(ValueError, match=f"`{method}` cannot be called before .*"):
+        tracker = MetricTracker(Accuracy(num_classes=10))
```
Comment (Borda) [inline tests/wrappers/test_tracker.py:41]:
```suggestion
    tracker = MetricTracker(Accuracy(num_classes=10))
    with pytest.raises(ValueError, match=f"`{method}` cannot be called before .*"):
```

**Review by Borda — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]
